src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderSetParent.cs

```diff
@@ -134,5 +134,29 @@ public void ParentOpenGenericClass_ThrowsBadImageFormatExceptionOnCreation()
 
             Assert.Throws<BadImageFormatException>(() => type.CreateTypeInfo());
         }
+
+        [Fact]
+        public static void RecursiveTypeBuilder_Self()
+        {
+            AssemblyBuilder assembly = Helpers.DynamicAssembly();
+            ModuleBuilder module = assembly.DefineDynamicModule("TestModule");
+
+            TypeBuilder type = module.DefineType("TestType", TypeAttributes.Class | TypeAttributes.Public);
+            type.SetParent(type);
+            Assert.Throws<ArgumentException>(() => type.CreateType());
+        }
+
+        [Fact]
+        public static void RecursiveTypeBuilder_Base()
+        {
+            AssemblyBuilder assembly = Helpers.DynamicAssembly();
+            ModuleBuilder module = assembly.DefineDynamicModule("TestModule");
+
+            TypeBuilder typeBase = module.DefineType("BaseType", TypeAttributes.Class | TypeAttributes.Public);
+            TypeBuilder typeDerived = module.DefineType("DerivedType", TypeAttributes.Class | TypeAttributes.Public);
+            typeBase.SetParent(typeDerived);
+            typeDerived.SetParent(typeBase);
+            Assert.Throws<ArgumentException>(() => typeBase.CreateType());
```

Comment:
I assume that the following variant of the repro:
```
ModuleBuilder module = assembly.DefineDynamicModule("TestModule");

TypeBuilder type = module.DefineType("TestType", TypeAttributes.Class | TypeAttributes.Public);
TypeBuilder type2 = module.DefineType("TestType2", TypeAttributes.Class | TypeAttributes.Public);
type.SetParent(type2);
type2.SetParent(type2);

type.CreateType();
```
is still broken. Is that correct?